src/appleseed/renderer/modeling/bsdf/energycompensation.cpp

```diff
@@ -0,0 +1,485 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "energycompensation.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/concepts/noncopyable.h"
+#include "foundation/image/color.h"
+#include "foundation/image/exrimagefilewriter.h"
+#include "foundation/image/image.h"
+#include "foundation/image/tile.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+
+// Boost headers.
+#include "boost/filesystem/fstream.hpp"
+
+// Standard headers.
+#include <algorithm>
+#include <iomanip>
+
+using namespace foundation;
+using namespace std;
+
+namespace bfs = boost::filesystem;
+
+namespace renderer
+{
+
+namespace
+{
+    void write_cpp_array(
+        const bfs::path& filename,
+        const char*      array_name,
+        const size_t     dir_array_size,
+        const size_t     avg_array_size,
+        const size_t     num_columns,
+        const float*     dir_table,
+        const float*     avg_table)
+    {
+        const size_t array_size = dir_array_size + avg_array_size;
+        assert(array_size % num_columns == 0);
+
+        bfs::ofstream of(filename);
+        of << setprecision(6) << fixed;
+
+        of << "extern const float " << array_name << "[" << array_size << "] = " << endl;
+        of << "{" << endl;
+        of << "    // Directional albedo." << endl;
+
+        size_t i = 0;
+        const float* p = dir_table;
+
+        while (i < dir_array_size)
+        {
+            of << "    ";
+
+            for (size_t j = 0; j < num_columns ; ++j)
+            {
+                of << *p++ << "f" << ", ";
+                ++i;
+            }
+
+            of << endl;
+        }
+
+        of << "    // Average albedo." << endl;
+
+        i = 0;
+        p = avg_table;
+        const size_t last = avg_array_size - 1;
+
+        while (i < avg_array_size)
+        {
+            of << "    ";
+
+            for (size_t j = 0; j < num_columns ; ++j)
+            {
+                of << *p++ << "f";
+                if (i != last) of << ", ";
+
+                ++i;
+            }
+
+            of << endl;
+        }
+
+
+        of << "};" << endl;
+    }
+}
+
+//
+// AlbedoTable3D class implementation.
+//
+
+AlbedoTable2D::AlbedoTable2D()
+  : TableSize(32)
+  , TableHeight(TableSize + 1)
+{
+    m_data = new float[array_size()];
+    m_albedo_table = m_data;
+    m_avg_table = m_albedo_table + TableSize * TableSize;
+}
+
+AlbedoTable2D::AlbedoTable2D(const float* table)
+  : TableSize(32)
+  , TableHeight(TableSize + 1)
+  , m_data(nullptr)
+{
+    m_albedo_table = const_cast<float*>(table);
```

Comment:
That `const_cast<>` and the one below are suspicious. Why have a `const float*` parameter then?

Reply:
The tables can be computed or can be initialized with a C++ array now. We compute them in unit tests and use the C++ array version for normal usage. In one case, I need the table to be const and in the other I need to modify the table. That's the reason of the const_cast<>